wal/src/message_queue_impl/helpers.rs

```diff
@@ -0,0 +1,221 @@
+};
+use crate::manager::RegionId;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "Failed to sync snapshot of region, region id:{}, topic:{}, err:{}",
+        region_id,
+        topic,
+        source
+    ))]
+    SyncSnapshot {
+        region_id: RegionId,
+        topic: String,
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display(
+        "Failed to clean logs of region, region id:{}, topic:{}, err:{}",
+        region_id,
+        topic,
+        source
+    ))]
+    CleanLogs {
+        region_id: RegionId,
+        topic: String,
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+}
+
+define_result!(Error);
+
+/// Encode the region snapshot and push to message queue
+///
+/// It will be locked before being called to keep the order of snapshots.
+pub struct SnapshotSynchronizer<Mq: MessageQueue> {
+    region_id: RegionId,
+    message_queue: Arc<Mq>,
+    meta_topic: String,
+    meta_encoding: MetaEncoding,
+}
+
+impl<Mq: MessageQueue> SnapshotSynchronizer<Mq> {
+    pub fn new(
+        region_id: RegionId,
+        message_queue: Arc<Mq>,
+        meta_topic: String,
+        meta_encoding: MetaEncoding,
+    ) -> Self {
+        info!(
+            "Snapshot synchronizer init, region id:{}, meta topic:{}",
+            region_id, meta_topic
+        );
+
+        Self {
+            region_id,
+            message_queue,
+            meta_topic,
+            meta_encoding,
+        }
+    }
+
+    pub async fn sync(&self, snapshot: RegionMetaSnapshot) -> Result<()> {
+        info!(
+            "Begin to sync snapshot to meta topic, snapshot:{:?}, topic:{}",
+            snapshot, self.meta_topic
+        );
+
+        let mut key_buf = BytesMut::new();
+        let mut value_buf = BytesMut::new();
+        self.meta_encoding
+            .encode_key(&mut key_buf, &MetaKey(self.region_id))
+            .map_err(|e| Box::new(e) as _)
+            .context(SyncSnapshot {
+                region_id: self.region_id,
+                topic: self.meta_topic.clone(),
+            })?;
+        self.meta_encoding
+            .encode_value(&mut value_buf, snapshot)
+            .map_err(|e| Box::new(e) as _)
+            .context(SyncSnapshot {
+                region_id: self.region_id,
+                topic: self.meta_topic.clone(),
+            })?;
+
+        // Try to write to message queue.
+        let messages = vec![to_message(key_buf.to_vec(), value_buf.to_vec())];
+        let offsets = self
+            .message_queue
+            .produce(&self.meta_topic, messages)
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(SyncSnapshot {
+                region_id: self.region_id,
+                topic: self.meta_topic.clone(),
+            })?;
+        assert_eq!(offsets.len(), 1);
+
+        // Delete old snapshots.
+        self.message_queue
+            .delete_up_to(&self.meta_topic, offsets[0])
```

Comment:
Is the offset passed to `delete_up_to`exclusive? It is weird that `up_to` seems inclusive if not. Is it wrong if it is.

Reply:
I modify `up_to` to `to` now.

Reply:
It seems no such modification now.

Reply:
I propose to make modification about message queue component together in a new pr, just keep it now?